### ISSUE TITLE
Wrong unused warning for type alias

### ISSUE THREAD

**Issue by therustmonk** [rust-lang/rust#59333]
This code:
```
struct Runner;

type RuntimeImpl = Runner;

trait Runtime {
    fn run(&mut self);
}

impl Runtime for &mut RuntimeImpl {
    fn run(&mut self) { }    
}

fn main() {
    let mut runner = Runner;
    (&mut runner).run();
}
```

takes the warning:
```
   Compiling playground v0.0.1 (/playground)
warning: type alias is never used: `RuntimeImpl`
 --> src/main.rs:3:1
  |
3 | type RuntimeImpl = Runner;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(dead_code)] on by default
```

But removing this alias introduces a compilation error :thinking: 

The link to the nightly playground example: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=22a044608f4d67680bcf1cc7fd23a381

**Comment by Lonami** [comment]
There is no compilation error if you remove the alias (and of course rename `impl ... for &mut Runner`), at least for me.

**Comment by Patryk27** [comment]
@Lonami: If you have to remove the alias _and_ rename the impl, it's wrong to say that the alias was not used in the first place - isn't it? It seems that Rust ignores fact that aliases can have impls and shows the `type alias is never used` warning anyway.

**Comment by Lonami** [comment]
@Patryk27 I'm not talking about the warning, which is what the issue is for (and it is a valid issue), I'm talking about:
> But removing this alias introduces a compilation error 🤔

**Comment by Patryk27** [comment]
If you remove the alias, there _is_ a compilation error: `cannot find type `RuntimeImpl` in this scope`. What @DenisKolodin meant, as I think, is that this error message contradicts the `type alias is never used` warning.

**Comment by DenisKolodin** [comment]
Yeah, I meant the alias is actually used and cannot be removed.

**Comment by taiki-e** [comment]
Related (or possibly the same) problem: #47131

**Comment by varkor** [comment]
So, https://github.com/rust-lang/rust/pull/59486 will fix *part* of this issue, which means that if `Runner` and `Runtime` are made `pub`, there will no longer be a dead code warning. However, there's a *second* issue, which is that the usages in code are not being picked up. I think this is because they're being attributed to the original type rather than the alias.

**Comment by varkor** [comment]
Minimised:
```rust
struct Runner;

type RuntimeImpl = Runner;

trait Runtime {
    fn run(&self) {}
}

impl Runtime for RuntimeImpl {}

fn main() {
    Runner.run();
}
```

**Comment by varkor** [comment]
A similar issue was encountered in https://github.com/rust-lang/rust/issues/47131#issuecomment-555712512:
```rust
trait Trait<S> {}

type Alias = ();

trait Trait2 {
    type Type: Trait<Alias>;
}
```
If these are not fixed at the same time, we should open a new issue for the unaddressed test case.

**Comment by Spoonbender** [comment]
triage: partially resolved

I'm not seeing errors for [the original post](https://github.com/rust-lang/rust/issues/59333#issue-423561062) nor for [this example](https://github.com/rust-lang/rust/issues/59333#issuecomment-477822694), but [2nd example](https://github.com/rust-lang/rust/issues/59333#issuecomment-555973113) still generates an incorrect warning.

**Comment by RodBurman** [comment]
I think this issue can be closed since the examples no longer give warnings when compiled. The 2nd example cited is for a different issue namely #47131.

**Comment by jakubadamw** [comment]
@RodBurman, that’s correct. Marking as E-needs-test.

@rustbot label +E-needs-test
